In [1]:
import pandas as pd

df = pd.read_csv(
    r"C:\Users\shyam\OneDrive\Documents\customer-churn-ltv-engine\data\raw\telco_churn.csv"
)

print(df.shape)

(7043, 33)


In [2]:
y = df['CLTV']

In [3]:
df = df.drop([
    'CustomerID',
    'Country',
    'State',
    'City',
    'Lat Long',
    'Churn Reason',
    'CLTV'
], axis=1)

In [4]:
df['Total Charges'] = pd.to_numeric(
    df['Total Charges'],
    errors='coerce'
)

df['Total Charges'] = df['Total Charges'].fillna(
    df['Total Charges'].median()
)

In [5]:
df_encoded = pd.get_dummies(df, drop_first=True)

print(df_encoded.shape)

(7043, 37)


In [6]:
X = df_encoded
y = pd.read_csv(
    r"C:\Users\shyam\OneDrive\Documents\customer-churn-ltv-engine\data\raw\telco_churn.csv"
)['CLTV']

print(X.shape)
print(y.shape)

(7043, 37)
(7043,)


In [7]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)

(5634, 37)
(1409, 37)


In [8]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train, y_train)

print("Model Trained Successfully")

Model Trained Successfully


In [9]:
y_pred = lr.predict(X_test)

print(y_pred[:5])

[4115.89762845 3991.8109072  3935.87530424 3889.97983717 4489.22968742]


In [10]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("MAE:", mae)
print("RMSE:", rmse)
print("R² Score:", r2)

MAE: 899.3175910843123
RMSE: 1063.7870970146705
R² Score: 0.16696410410197327


In [11]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

print("Random Forest Model Trained")

Random Forest Model Trained


In [12]:
y_pred_rf = rf.predict(X_test)

In [13]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

mae_rf = mean_absolute_error(y_test, y_pred_rf)
rmse_rf = np.sqrt(mean_squared_error(y_test, y_pred_rf))
r2_rf = r2_score(y_test, y_pred_rf)

print("MAE:", mae_rf)
print("RMSE:", rmse_rf)
print("R² Score:", r2_rf)

MAE: 893.1562384669978
RMSE: 1055.1049607966972
R² Score: 0.18050631919914406


## Final Conclusion

Two regression models were evaluated for CLTV prediction.

1. Linear Regression achieved an R² score of 0.167.
2. Random Forest Regressor achieved an R² score of 0.181.
3. Random Forest performed better across all evaluation metrics.
4. The model can be used to identify high-value customers and support retention strategies.

Selected Model: Random Forest Regressor